src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -293,53 +357,52 @@ private ClassDesc getNearestClassWithSameName(String className, String superClas
+        if (shouldSearchAnonSuperClass) {
+            diff = Integer.compare(
+                CheckUtil
+                    .typeDeclarationMatchingCountAnonSuperClass(superOrOuterClassName,
+                                                                secondClass.getQualifiedName()),
+                CheckUtil
+                    .typeDeclarationMatchingCountAnonSuperClass(superOrOuterClassName,
+                                                                firstClass.getQualifiedName()));
+        }
+        else {
+            diff = Integer.compare(
+                CheckUtil.typeDeclarationNameMatchingCount(superOrOuterClassName,
+                                                           secondClass.getQualifiedName()),
+                CheckUtil.typeDeclarationNameMatchingCount(superOrOuterClassName,
+                                                           firstClass.getQualifiedName()));
+        }
         if (diff == 0) {
             diff = Integer.compare(firstClass.getDepth(), secondClass.getDepth());
         }
         return diff;
     }
 
     /**
-     * Check if class name matches with anonymous inner class name.
-     *
-     * @param ast current ast.
-     * @param classDesc class to match.
-     * @return true if current class name matches anonymous inner
-     *         class name.
-     */
-    private static boolean doesNameOfClassMatchAnonymousInnerClassName(DetailAST ast,
-                                                               ClassDesc classDesc) {
-        final String[] className = classDesc.getQualifiedName().split("\\.");
-        return ast.getFirstChild().getText().equals(className[className.length - 1]);
-    }
-
-    /**
-     * Get qualified class name from given class Ast.
+     * Get qualified type declaration name from given type declaration Ast.
      *
-     * @param classAst class to get qualified class name
-     * @return qualified class name of a class
+     * @param typeDeclarationAst type declaration for which qualified name is being fetched
+     * @return qualified name of a type declaration
      */
-    private String getQualifiedClassName(DetailAST classAst) {
-        final String className = classAst.findFirstToken(TokenTypes.IDENT).getText();
-        String outerClassQualifiedName = null;
-        if (!classes.isEmpty()) {
-            outerClassQualifiedName = classes.peek().getQualifiedName();
+    private String getQualifiedTypeDeclarationName(DetailAST typeDeclarationAst) {
+        final String className = typeDeclarationAst.findFirstToken(TokenTypes.IDENT).getText();
+        String outerTypeDeclarationQualifiedName = null;
+        if (!typeDeclarations.isEmpty()) {
+            outerTypeDeclarationQualifiedName = typeDeclarations.peek().getQualifiedName();
         }
-        return CheckUtil
-            .getQualifiedTypeDeclarationName(packageName, outerClassQualifiedName, className);
+        return CheckUtil.getQualifiedTypeDeclarationName(packageName,
```

Comment:
It is intentional, the complete name is built in 2 steps, one method is stateful whereas the other one is stateless.

>Looks confusing

We could refactor it to eliminate the stateful method and do it in the place where the method is called, but I feel this type of method is pretty common, a method calling another method with the same name with different arguments.

Reply:
> method calling another method with the same name with different arguments.

this is ok in case of overloading. But it seems confusing to me and can be error-prone (e.g. if you just search for method by text search). I suggest to rename method in check, can be `extractQualifiedTypeName`

Reply:
Done.